sys/vm/vm_cheri_revoke.c

```diff
@@ -56,10 +63,161 @@
+
+static COUNTER_U64_DEFINE_EARLY(cheri_revoke_errors);
+SYSCTL_COUNTER_U64(_vm_stats_cheri_revoke, OID_AUTO, errors, CTLFLAG_RD,
+    &cheri_revoke_errors,
+    "Number of error returns from the page fault handler");
+
+static COUNTER_U64_DEFINE_EARLY(cheri_revoke_restarts);
+SYSCTL_COUNTER_U64(_vm_stats_cheri_revoke, OID_AUTO, restarts, CTLFLAG_RD,
+    &cheri_revoke_restarts,
+    "Number of VM map re-lookups");
+
+/***************************** KERNEL THREADS ***************************/
+
+static MALLOC_DEFINE(M_REVOKE, "cheri_revoke", "cheri_revoke temporary data");
+
+struct async_revoke_cookie {
+	struct vmspace	*vm;
+	struct vm_cheri_revoke_cookie cookie;
+	TAILQ_ENTRY(async_revoke_cookie) link;
+};
+
+static struct mtx async_revoke_mtx;
+MTX_SYSINIT(async_revoke_mtx, &async_revoke_mtx, "async revoke", MTX_DEF);
+static TAILQ_HEAD(, async_revoke_cookie) async_revoke_work =
+    TAILQ_HEAD_INITIALIZER(async_revoke_work);
+
+static void
+vm_cheri_revoke_pass_async_pre(vm_map_t map, struct vm_cheri_revoke_cookie *crc)
+{
+	cheri_revoke_epoch_t epoch;
+	enum cheri_revoke_state state __diagused;
+
+	vm_map_lock(map);
+	epoch = cheri_revoke_st_get_epoch(map->vm_cheri_async_revoke_st);
+	state = cheri_revoke_st_get_state(map->vm_cheri_async_revoke_st);
+
+	KASSERT((epoch & 1) == 0,
+	    ("unexpected async revoke epoch %lu", epoch));
+	KASSERT(state == CHERI_REVOKE_ST_INITING,
+	    ("unexpected async revoke state %d", state));
+
+	cheri_revoke_st_set(&map->vm_cheri_async_revoke_st, epoch + 1,
+	    CHERI_REVOKE_ST_INITED);
+	map->vm_cheri_async_revoke_shadow = crc->crshadow;
+	vm_map_unlock(map);
+}
+
+static void
+vm_cheri_revoke_pass_async_post(vm_map_t map, int error)
+{
+	cheri_revoke_epoch_t epoch;
+	enum cheri_revoke_state state __diagused;
+
+	vm_map_lock(map);
+	epoch = cheri_revoke_st_get_epoch(map->vm_cheri_async_revoke_st);
+	state = cheri_revoke_st_get_state(map->vm_cheri_async_revoke_st);
+
+	KASSERT((epoch & 1) == 1,
+	    ("unexpected post-async revoke epoch %lu", epoch));
+	KASSERT(state == CHERI_REVOKE_ST_INITED,
+	    ("unexpected post-async revoke state %d", state));
+
+	map->vm_cheri_async_revoke_status = error;
+	if (error == KERN_SUCCESS)
+		epoch++;
+	cheri_revoke_st_set(&map->vm_cheri_async_revoke_st, epoch,
+	    CHERI_REVOKE_ST_CLOSING);
+	map->vm_cheri_async_revoke_shadow = NULL;
+	vm_map_unlock(map);
+}
+
+static void
+vm_cheri_revoke_kproc(void *arg __unused)
+{
+	struct vmspace *myvm;
+
+	/*
+	 * Make sure that our own vmspace doesn't go away when we switch.
+	 */
+	myvm = vmspace_acquire_ref(curproc);
+
+	for (;;) {
+		struct async_revoke_cookie *arc;
+		vm_map_t map;
+		int error;
+
+		mtx_lock(&async_revoke_mtx);
+		if ((arc = TAILQ_FIRST(&async_revoke_work)) == NULL) {
+			if (curproc->p_vmspace != myvm) {
+				mtx_unlock(&async_revoke_mtx);
+				vmspace_switch_aio(myvm);
+				continue;
+			}
+			msleep(&async_revoke_work, &async_revoke_mtx, 0, "rvkslp",
+			    0);
+			mtx_unlock(&async_revoke_mtx);
```

Comment:
It is absolutely a nitpick, and fine to ignore, but: having `msleep` acquire the lock only for us to drop it and then immediately reacquire it is very slightly sad.

Reply:
There is a `PDROP` flag (not often used) that doesn't re-acquire the lock on wakeup.

Reply:
Fixed, this is an artifact of having structured the loop differently when I first wrote it.